File: factory/kassabild.py
#!/usr/bin/env python3
"""kassabild.py — den SPRÅKLÖSA trygghetsbilden till Shopifys kassa.

    python3 factory/kassabild.py --spec <spec.json> --ut <bild.png>

Varför den är språklös, och varför det inte är en stilfråga:

Shopifys kassa går INTE att anpassa per marknad under planen Advanced
(mätt på CaraShell 2026-09-21: planen är "Shopify", checkoutBranding svarar
ACCESS_DENIED med plantexten, och butiken har EN publicerad checkout-profil
som betjänar alla fem marknaderna). Bilden är alltså EN inställning för hela
butiken, och den visas samtidigt för svenska, norska, danska, finska och
engelsktalande kunder. Ett enda ord på bilden är därför fel för fyra av fem.

Det som ÄR språkneutralt: butikens eget namn (samma i alla marknader),
stjärnor, och en siffra. Inget annat får ritas här — `granska_ord()` vägrar
skriva filen om spec:en bär text som inte är butiksnamnet eller ett tal.

⚠️ Bilden är STATISK. Betyget måste därför mätas, inte skrivas: kassabild.mjs
läser produkternas `reviews.rating` live ur Shopify och stoppar in det här,
och bokför vad den mätte så en senare körning kan se att siffran glidit.

Stjärnorna ritas som polygoner, aldrig som tecknet ★ — Liberation Sans (det
enda typsnitt containern har med garanterat å/ä/ö) saknar U+2605, och ett
tecken som saknas blir en tom ruta utan felmeddelande.

spec.json:
    {
      "namn": "CaraShell",
      "betyg": 5.0,
      "farger": { "mork": "#22282E", "accent": "#1F6F8E", "text": "#1B2026",
                  "yta": "#FFFFFF", "stjarna": "#F5A623" },
      "logga": "valfri/sokvag/logga.png",
      "bredd": 560
    }
"""
import argparse
import json
import math
import os
import re
import sys

try:
    from PIL import Image, ImageDraw, ImageFont
except ImportError:
    sys.exit('Pillow saknas. Kör: pip3 install pillow')

TYPSNITT = [
    '/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf',
    '/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf',
]


def typsnitt(px):
    for s in TYPSNITT:
        if os.path.exists(s):
            return ImageFont.truetype(s, px)
    raise SystemExit(
        'Hittade inget typsnitt med å/ä/ö. Sökta: ' + ', '.join(TYPSNITT) +
        '. Installera fonts-liberation — ett saknat typsnitt får aldrig bli '
        'ett tyst fallback till ett utan svenska tecken.'
    )


def granska_ord(spec):
    """Vägrar varje text som inte är butiksnamnet eller ett tal.

    Spärren finns för att bilden visas i FEM marknader samtidigt. Ett
    "Trygg betalning" här är osynligt fel för den danska kunden — det ser
    rätt ut i förhandsgranskningen och är fel i fyra av fem kassor.
    """
    namn = str(spec.get('namn', '')).strip()
    fel = []
    for nyckel in ('rubrik', 'underrad', 'text', 'badge', 'botten'):
        if str(spec.get(nyckel, '')).strip():
            fel.append(f'{nyckel}: "{spec[nyckel]}"')
    if fel:
        raise SystemExit(
            'STOPP — kassabilden får inte bära ord. Den visas i alla marknader '
            'samtidigt (Shopify kan inte byta den per marknad under planen '
            'Advanced). Ta bort: ' + '; '.join(fel)
        )
    if namn and not re.fullmatch(r'[A-Za-zÅÄÖåäö0-9 .&-]{1,24}', namn):
        raise SystemExit(f'STOPP — "{namn}" ser inte ut som ett butiksnamn.')
    return namn


def stjarna(rita, mitt, radie, farg, fyllnad=1.0):
    """En femuddig stjärna som polygon. `fyllnad` 0–1 klipper den från vänster."""
    cx, cy = mitt
    punkter = []
    for i in range(10):
        r = radie if i % 2 == 0 else radie * 0.42
        v = math.radians(-90 + i * 36)
        punkter.append((cx + r * math.cos(v), cy + r * math.sin(v)))
    if fyllnad >= 0.999:
        rita.polygon(punkter, fill=farg)
        return
    # Delvis fylld: rita hela i blekt och klipp in den fyllda delen.
    rita.polygon(punkter, fill=farg + (60,) if len(farg) == 3 else farg)
    if fyllnad <= 0.001:
        return
    lager = Image.new('RGBA', (int(radie * 4), int(radie * 4)), (0, 0, 0, 0))
    d2 = ImageDraw.Draw(lager)
    flytt = [(x - cx + radie * 2, y - cy + radie * 2) for x, y in punkter]
    d2.polygon(flytt, fill=farg)
    bredd = int(radie * 4 * fyllnad)
    beskuren = lager.crop((0, 0, bredd, int(radie * 4)))
    lager2 = Image.new('RGBA', lager.size, (0, 0, 0, 0))
    lager2.paste(beskuren, (0, 0))
    rita._image.alpha_composite(lager2, (int(cx - radie * 2), int(cy - radie * 2)))


def hex_rgb(v, standard=(0, 0, 0)):
    s = str(v or '').strip().lstrip('#')
    if len(s) != 6:
        return standard
    try:
        return tuple(int(s[i:i + 2], 16) for i in (0, 2, 4))
    except ValueError:
        return standard


def bygg(spec, ut):
    namn = granska_ord(spec)
    betyg = float(spec.get('betyg') or 0)
    if not 0 < betyg <= 5:
        raise SystemExit(f'STOPP — betyget {betyg} är inte mätt. Kör kassabild.mjs, som läser det ur Shopify.')
    f = spec.get('farger') or {}
    c_text = hex_rgb(f.get('mork'), (34, 40, 46))
    c_stjarna = hex_rgb(f.get('stjarna') or '#F5A623', (245, 166, 35))
    c_yta = hex_rgb(f.get('yta') or '#FFFFFF', (255, 255, 255))

    bredd = int(spec.get('bredd') or 560)
    hojd = int(bredd * 0.26)
    skala = 4  # ritas 4× och krymps — kanterna blir mjuka utan antialias-flaggor
    B, H = bredd * skala, hojd * skala
    bild = Image.new('RGBA', (B, H), c_yta + (255,))
    rita = ImageDraw.Draw(bild)
    rita._image = bild

    logga_sokvag = str(spec.get('logga') or '').strip()
    x = int(B * 0.06)
    mitt_y = H // 2

    if logga_sokvag and os.path.exists(logga_sokvag):
        logga = Image.open(logga_sokvag).convert('RGBA')
        mal_h = int(H * 0.46)
        mal_b = max(1, int(logga.width * mal_h / logga.height))
        logga = logga.resize((mal_b, mal_h), Image.LANCZOS)
        bild.alpha_composite(logga, (x, mitt_y - mal_h // 2))
        x += mal_b
    elif namn:
        tf = typsnitt(int(H * 0.30))
        rita.text((x, mitt_y), namn, font=tf, fill=c_text + (255,), anchor='lm')
        x += int(rita.textlength(namn, font=tf))

    # Avdelare mellan brandet och betyget.
    delare_x = x + int(B * 0.045)
    rita.line([(delare_x, int(H * 0.28)), (delare_x, int(H * 0.72))],
              fill=hex_rgb(f.get('linje') or '#D6D9D3', (214, 217, 211)) + (255,), width=max(1, skala))

    # Siffran — ett tal är samma på alla fem språk. Decimaltecknet är det enda
    # som skiljer (4.8 mot 4,8), så heltal skrivs utan decimal och resten med
    # den svenska kommateringen: kassans största marknad är Sverige.
    txt = str(int(betyg)) if betyg == int(betyg) else f'{betyg:.1f}'.replace('.', ',')
    tf = typsnitt(int(H * 0.26))
    txt_b = rita.textlength(txt, font=tf)

    # ⚠️ Stjärnornas storlek RÄKNAS ut ur det som är kvar, den sätts inte.
    # Med en fast radie sköt femte stjärnan och siffran utanför kanten, och
    # en bild som är beskuren i högerkant syns inte förrän någon tittar i en
    # riktig kassa — PNG:en själv ser hel ut. Ett långt butiksnamn, ett
    # bredare betyg (4,6 mot 5) eller en annan bredd flyttar gränsen.
    hoger = int(B * 0.94)
    sx = delare_x + int(B * 0.045)
    lucka = int(B * 0.025)
    tillgangligt = hoger - sx - lucka - txt_b
    radie = max(4, min(int(H * 0.17), int(tillgangligt / 11.75)))
    steg = int(radie * 2.35)
    if tillgangligt <= 0:
        raise SystemExit('STOPP — butiksnamnet tar hela bredden, stjärnorna får inte plats. Höj "bredd" i spec:en.')

    for i in range(5):
        kvar = betyg - i
        fyllnad = 1.0 if kvar >= 1 else (kvar if kvar > 0 else 0.0)
        stjarna(rita, (sx + radie + i * steg, mitt_y - int(H * 0.02)), radie, c_stjarna, fyllnad)

    rita.text((sx + 5 * steg + lucka, mitt_y - int(H * 0.02)), txt,
              font=tf, fill=c_text + (255,), anchor='lm')

    bild = bild.resize((bredd, hojd), Image.LANCZOS)
    os.makedirs(os.path.dirname(os.path.abspath(ut)) or '.', exist_ok=True)
    bild.save(ut)
    return {'fil': ut, 'bredd': bredd, 'hojd': hojd, 'betyg': betyg, 'namn': namn}


def main():
    p = argparse.ArgumentParser()
    p.add_argument('--spec', required=True)
    p.add_argument('--ut', required=True)
    p.add_argument('--json', action='store_true')
    a = p.parse_args()
    with open(a.spec, encoding='utf-8') as fh:
        spec = json.load(fh)
    res = bygg(spec, a.ut)
    if a.json:
        print(json.dumps(res, ensure_ascii=False))
    else:
        print(f"✅ {res['fil']} — {res['bredd']}×{res['hojd']} px, betyg {res['betyg']}, inga ord")


if __name__ == '__main__':
    main()

File: factory/kassabild.mjs
// kassabild.mjs — den språklösa trygghetsbilden till Shopifys kassa.
//
//   node factory/kassabild.mjs <butik> [--ut <fil.png>] [--bredd 560] [--torr]
//   node factory/kassabild.mjs <butik> --kolla     # har betyget glidit?
//
// Mäter butikens RIKTIGA betyg ur produkternas `reviews.rating` (Judge.me
// håller metafältet uppdaterat), renderar strippen med factory/kassabild.py
// och laddar upp den till butikens Files. Bokför vad den mätte i
// factory/output/<butik>/kassabild.json.
//
// ── Varför den finns, och varför den ser ut som den gör ──────────────────
//
// Axels fråga 2026-09-21: "det går att lägga till en bild i kassan … hur ska
// vi komma runt i alla marknader?" Mätningen samma dag gav svaret:
//
//   • CaraShell ligger på planen "Shopify". Att anpassa kassan PER MARKNAD
//     kräver Advanced eller Plus — butiken är ett steg under gränsen.
//   • Checkout Branding API:t svarar ACCESS_DENIED med plantexten, trots att
//     appen har write_checkout_branding_settings. Scope och plan är två grindar.
//   • Butiken har EN publicerad checkout-profil för alla fem marknaderna.
//   • `TranslatableResourceType` har ingen checkout-resurs: kassans utseende
//     går inte att översätta.
//   • Shopify tog dessutom bort bakgrundsbilden i Header och Main 2026-02-05,
//     på alla planer. Kvar: logotypen och ordersammanfattningens bakgrund —
//     och ordersammanfattningen är HOPFÄLLD som standard på mobil, där i
//     stort sett all Meta-trafik landar.
//
// Slutsatsen: bilden blir EN bild för fem marknader, och den enda ytan som
// syns på mobil är logotypsfältet. Därför: en strip UTAN ord (butiksnamn,
// stjärnor, en siffra), som fungerar identiskt på svenska, norska, danska,
// finska och engelska — och som behåller brandet i kassan i stället för att
// byta ut det. Spärren mot ord sitter i kassabild.py, inte här.
//
// ⚠️ SISTA STEGET ÄR EN MÄNNISKA. API:t är stängt av planen, så bilden måste
// väljas i admin: Inställningar → Kassa → Anpassa → kugghjulet → Logo →
// Image → Select from library. Skriptet skriver ut exakt den raden.
//
// ⚠️ BILDEN ÄR STATISK. Betyget som mäts i dag är inte betyget om en månad.
// `--kolla` mäter om och säger till när siffran glidit så mycket att bilden
// ljuger; den skriver inget och laddar inte upp något.

import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { spawnSync } from 'node:child_process';
import { dirname, join } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { lasYaml } from './yaml.mjs';
import { graphql } from './shopify.mjs';
import { anslut } from './token.mjs';
import { laddaUppBild } from './filer.mjs';
import { hamtaArbetstema, hamtaTemafil } from './shopify.mjs';
import { lasTemaJson } from './tema.mjs';

const ROT = join(dirname(fileURLToPath(import.meta.url)), '..');

/** Så mycket får betyget glida innan bilden räknas som osann. */
export const DRIFTGRANS = 0.2;

/**
 * Butikens betyg ur produkternas metafält, viktat på antalet recensioner.
 *
 * Viktat, inte medel av medel: en produkt med 4 recensioner ska inte väga
 * lika tungt som en med 200. Produkter utan recensioner hoppas helt — en
 * nolla hade dragit ner snittet till något som aldrig stått någonstans.
 */
export function vagtBetyg(produkter) {
  const med = (produkter ?? []).filter((p) => Number(p?.antal) > 0 && Number(p?.betyg) > 0);
  if (med.length === 0) return null;
  const antal = med.reduce((s, p) => s + Number(p.antal), 0);
  const summa = med.reduce((s, p) => s + Number(p.betyg) * Number(p.antal), 0);
  // ⚠️ `antalProdukter`, inte `produkter`: fältet spreadas ihop med
  // produktLISTAN i anroparen, och med samma namn skrev listan över antalet —
  // rapporten skrev "[object Object],[object Object] produkter" (mätt 2026-09-21).
  return { betyg: Math.round((summa / antal) * 10) / 10, antal, antalProdukter: med.length };
}

/** Läser betyg + antal per produkt ur butiken. Bara läsningar. */
export async function hamtaBetyg() {
  const d = await graphql(`{
    products(first: 50, query: "status:active") {
      nodes {
        title handle
        betyg: metafield(namespace: "reviews", key: "rating") { value }
        antal: metafield(namespace: "reviews", key: "rating_count") { value }
      }
    }
  }`);
  return (d.products?.nodes ?? []).map((p) => {
    let betyg = null;
    try { betyg = Number(JSON.parse(p.betyg?.value ?? 'null')?.value ?? NaN); } catch { betyg = null; }
    return { titel: p.title, handle: p.handle, betyg: Number.isFinite(betyg) ? betyg : null, antal: Number(p.antal?.value ?? 0) || 0 };
  });
}

/**
 * Butikens riktiga logga: temats `settings.logo` (shopify://shop_images/<fil>)
 * → filens CDN-adress → hämtad till disk. Utan den ritar kassabild.py
 * butiksnamnet i Liberation Sans, och det dömde Axel ut 2026-09-21.
 * Returnerar null när temat saknar logga — då säger rapporten det.
 */
export async function hamtaLogga(utMapp) {
  const tema = await hamtaArbetstema();
  const ra = await hamtaTemafil(tema.id, 'config/settings_data.json');
  const logo = String(lasTemaJson(ra ?? '{}')?.current?.logo ?? '');
  const m = logo.match(/shop_images\/([^?]+)/);
  if (!m) return null;
  const filnamn = m[1];
  const stam = filnamn.replace(/\.[a-z0-9]+$/i, '');
  const d = await graphql(`{ files(first: 10, query: "filename:${stam}") { nodes { ... on MediaImage { image { url } } } } }`);
  const traff = (d.files?.nodes ?? []).map((n) => n.image?.url).find((u) => u && u.includes(`/${filnamn}`));
  if (!traff) return null;
  const r = await fetch(traff);
  if (!r.ok) return null;
  mkdirSync(utMapp, { recursive: true });
  const lokal = join(utMapp, `logga-${filnamn}`);
  writeFileSync(lokal, Buffer.from(await r.arrayBuffer()));
  return { filnamn, url: traff, lokal };
}

const bokforing = (butikId) => join(ROT, 'factory', 'output', butikId, 'kassabild.json');

export function lasBokforing(butikId) {
  const f = bokforing(butikId);
  if (!existsSync(f)) return null;
  try { return JSON.parse(readFileSync(f, 'utf8')); } catch { return null; }
}

/** Har det uppmätta betyget glidit så mycket att den uppladdade bilden ljuger? */
export function drift(bokfort, nu) {
  if (!bokfort?.betyg || !nu?.betyg) return { glidit: false, orsak: 'ingen tidigare mätning' };
  const d = Math.round(Math.abs(Number(nu.betyg) - Number(bokfort.betyg)) * 10) / 10;
  return {
    glidit: d > DRIFTGRANS,
    diff: d,
    orsak: d > DRIFTGRANS
      ? `bilden visar ${bokfort.betyg}, butiken har ${nu.betyg} — bygg om den och byt i admin`
      : `bilden visar ${bokfort.betyg}, butiken har ${nu.betyg}`,
  };
}

function rita(spec, ut) {
  mkdirSync(dirname(ut), { recursive: true });
  const specfil = `${ut}.spec.json`;
  writeFileSync(specfil, `${JSON.stringify(spec, null, 2)}\n`);
  const r = spawnSync('python3', [join(ROT, 'factory', 'kassabild.py'), '--spec', specfil, '--ut', ut, '--json'], { encoding: 'utf8' });
  if (r.status !== 0) throw new Error(`kassabild.py misslyckades:\n${(r.stderr || r.stdout || '').trim()}`);
  return JSON.parse(r.stdout.trim().split('\n').pop());
}

export async function byggKassabild(butikId, { bredd = 560, ut = null, torr = false, kolla = false } = {}) {
  const butiksfil = join(ROT, 'factory', 'butiker', `${butikId}.yaml`);
  if (!existsSync(butiksfil)) throw new Error(`Hittar ingen butiksfil: ${butiksfil}`);
  const butik = lasYaml(readFileSync(butiksfil, 'utf8'));

  // ⚠️ Adressen ur BUTIKSFILEN, aldrig ur miljöns SHOPIFY_SHOP. Utan
  // `onskadDoman` löser anslut() butiks-id:t mot vilken butik miljön råkar
  // peka på — mätt 2026-09-21: `kassabild.mjs carashell` landade på
  // TankGuards butik och stoppades först av state-spärren. Butiken ÄR
  // argumentet (KEDJAN.md), och `butik.myshopify` är facit.
  const myshopify = String(butik?.butik?.myshopify ?? '').trim();
  if (!myshopify) throw new Error(`${butiksfil} saknar butik.myshopify — utan den går det inte att veta vilken butik som menas.`);
  await anslut(butikId, { utanEnvFil: true, kravScopes: false, onskadDoman: myshopify });
  const produkter = await hamtaBetyg();
  const vagt = vagtBetyg(produkter);
  if (!vagt) {
    return { butikId, betyg: null, produkterRad: produkter, stopp: 'Ingen produkt har recensioner än — det finns inget betyg att visa. Importera recensionerna i Judge.me först.' };
  }

  const bokfort = lasBokforing(butikId);
  const d = drift(bokfort, vagt);
  if (kolla) return { butikId, ...vagt, produkterRad: produkter, bokfort, drift: d, kolla: true };

  const namn = String(butik?.butik?.brand ?? butikId);
  const f = butik?.branding?.farger ?? {};
  const fil = ut ?? join(ROT, 'factory', 'output', butikId, 'kassabild.png');
  const logga = await hamtaLogga(dirname(fil));
  const spec = {
    namn,
    betyg: vagt.betyg,
    bredd,
    logga: logga?.lokal ?? '',
    farger: { mork: f.mork, accent: f.accent, yta: 'transparent', linje: f.linje, stjarna: '#F5A623' },
  };
  const ritad = rita(spec, fil);

  if (torr) return { butikId, ...vagt, produkterRad: produkter, spec, bild: ritad, logga, drift: d, torr: true };

  const uppladdad = await laddaUppBild(fil, { alt: `${namn} ${vagt.betyg}`, filnamn: `${butikId}-kassabild.png`, aterandvand: false });
  const post = { butikId, namn, betyg: vagt.betyg, antal: vagt.antal, antalProdukter: vagt.antalProdukter, bild: ritad, logga: logga?.filnamn ?? null, fil: uppladdad, matt: new Date().toISOString().slice(0, 10) };
  mkdirSync(dirname(bokforing(butikId)), { recursive: true });
  writeFileSync(bokforing(butikId), `${JSON.stringify(post, null, 2)}\n`);
  return { ...post, produkterRad: produkter, drift: d };
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? '').href) {
  const argv = process.argv.slice(2);
  const butikId = argv.find((a) => !a.startsWith('--'));
  const flagga = (n) => argv.includes(`--${n}`);
  const varde = (n) => { const i = argv.indexOf(`--${n}`); return i >= 0 ? argv[i + 1] : null; };
  if (!butikId) {
    console.error('Användning: node factory/kassabild.mjs <butik> [--ut fil.png] [--bredd 560] [--torr|--kolla]');
    process.exit(2);
  }
  try {
    const r = await byggKassabild(butikId, {
      bredd: Number(varde('bredd')) || 560,
      ut: varde('ut'),
      torr: flagga('torr'),
      kolla: flagga('kolla'),
    });
    if (r.stopp) { console.error(`⛔ ${r.stopp}`); process.exit(1); }
    console.log(`Betyg per produkt (ur reviews.rating, Judge.me):`);
    for (const p of (r.produkterRad ?? [])) {
      console.log(`  ${p.antal > 0 ? '✅' : '⏭ '} ${p.titel?.slice(0, 44).padEnd(44)} ${p.betyg ?? '–'} på ${p.antal} recensioner`);
    }
    console.log(`\nViktat betyg: ${r.betyg} på ${r.antal} recensioner över ${r.antalProdukter} produkter`);
    if (r.drift?.orsak) console.log(`Drift: ${r.drift.glidit ? '⚠️ ' : ''}${r.drift.orsak}`);
    if (r.kolla) process.exit(r.drift?.glidit ? 1 : 0);
    console.log(`Logga: ${r.logga?.filnamn ?? r.logga ?? '⚠️ SAKNAS i temat — butiksnamnet ritas som text i stället'}`);
    console.log(`Bild: ${r.bild.fil} (${r.bild.bredd}×${r.bild.hojd} px, inga ord)`);
    if (r.torr) { console.log('\nTorrkörning — inget laddades upp.'); process.exit(0); }
    console.log(`Uppladdad: ${r.fil?.url ?? r.fil?.id ?? '(okänt)'}`);
    console.log('\n🖐 SISTA STEGET GÖRS AV EN MÄNNISKA (API:t är stängt av planen):');
    console.log('   1. Shopify admin → Inställningar → Kassa');
    console.log('   2. Klicka Anpassa');
    console.log('   3. Klicka kugghjulet uppe till höger');
    console.log('   4. Under Logo: klicka Image → Select → Select from library');
    console.log(`   5. Välj filen ${butikId}-kassabild.png och klicka Spara`);
  } catch (e) {
    console.error(`❌ ${e.message}`);
    process.exit(1);
  }
}
